File: apps/mobile/ios/Sources/Models/ComponentState.swift
//
//  ComponentState.swift
//  TchatApp
//
//  Created by Claude on 22/09/2024.
//

import Foundation
import SwiftUI

/// Component state entity for UI component state management and synchronization
public struct ComponentState: Codable, Identifiable {

    // MARK: - Properties

    public let id: String
    public let componentId: String
    public let instanceId: String
    public let userId: String
    public let sessionId: String
    public var state: [String: Any]
    public let timestamp: Date
    public var version: Int
    public let platform: String
    public var metadata: ComponentStateMetadata

    // MARK: - Initialization

    public init(
        id: String = UUID().uuidString,
        componentId: String,
        instanceId: String,
        userId: String,
        sessionId: String,
        state: [String: Any] = [:],
        timestamp: Date = Date(),
        version: Int = 1,
        platform: String = "ios",
        metadata: ComponentStateMetadata = ComponentStateMetadata()
    ) {
        self.id = id
        self.componentId = componentId
        self.instanceId = instanceId
        self.userId = userId
        self.sessionId = sessionId
        self.state = state
        self.timestamp = timestamp
        self.version = version
        self.platform = platform
        self.metadata = metadata
    }

    // MARK: - Computed Properties

    /// Check if state is synchronized
    public var isSynchronized: Bool {
        return metadata.lastSyncTimestamp != nil &&
               Date().timeIntervalSince(metadata.lastSyncTimestamp!) < 10.0 // 10 seconds
    }

    /// Get state keys
    public var stateKeys: [String] {
        return Array(state.keys)
    }

    /// Check if state is empty
    public var isEmpty: Bool {
        return state.isEmpty
    }

    /// Get state size (number of properties)
    public var stateSize: Int {
        return state.count
    }

    /// Calculate state hash for conflict detection
    public var stateHash: String {
        let sortedKeys = state.keys.sorted()
        let stateString = sortedKeys.map { "\($0):\(state[$0] ?? "")" }.joined(separator: "|")
        return String(stateString.hashValue)
    }

    // MARK: - State Operations

    /// Update specific state property
    public mutating func updateProperty(_ key: String, value: Any) {
        state[key] = value
        incrementVersion()
    }

    /// Update multiple state properties
    public mutating func updateProperties(_ updates: [String: Any]) {
        for (key, value) in updates {
            state[key] = value
        }
        incrementVersion()
    }

    /// Remove state property
    public mutating func removeProperty(_ key: String) {
        state.removeValue(forKey: key)
        incrementVersion()
    }

    /// Clear all state
    public mutating func clearState() {
        state.removeAll()
        incrementVersion()
    }

    /// Merge state with another state
    public mutating func mergeState(_ otherState: [String: Any], strategy: MergeStrategy = .overwrite) {
        switch strategy {
        case .overwrite:
            for (key, value) in otherState {
                state[key] = value
            }
        case .keepExisting:
            for (key, value) in otherState {
                if state[key] == nil {
                    state[key] = value
                }
            }
        case .merge:
            for (key, value) in otherState {
                state[key] = value
            }
        }
        incrementVersion()
    }

    /// Get state property with type safety
    public func getProperty<T>(_ key: String, as type: T.Type) -> T? {
        return state[key] as? T
    }

    /// Get state property with default value
    public func getProperty<T>(_ key: String, default defaultValue: T) -> T {
        return state[key] as? T ?? defaultValue
    }

    // MARK: - Sync Operations

    /// Create state sync request
    public func createSyncRequest() -> ComponentStateSyncRequest {
        return ComponentStateSyncRequest(
            userId: userId,
            sessionId: sessionId,
            platform: platform,
            componentStates: [self],
            timestamp: Date(),
            syncVersion: version
        )
    }

    /// Apply sync response
    public mutating func applySyncResponse(_ response: ComponentStateSyncResponse) {
        if response.success {
            self.version = response.syncVersion
            markSynchronized()
        }
    }

    /// Mark state as synchronized
    public mutating func markSynchronized() {
        metadata = ComponentStateMetadata(
            lastSyncTimestamp: Date(),
            syncAttempts: metadata.syncAttempts,
            conflictCount: metadata.conflictCount,
            customData: metadata.customData
        )
    }

    /// Increment version
    private mutating func incrementVersion() {
        version += 1
    }

    /// Detect conflicts with another state
    public func hasConflictWith(_ otherState: ComponentState) -> Bool {
        return componentId == otherState.componentId &&
               instanceId == otherState.instanceId &&
               version != otherState.version &&
               stateHash != otherState.stateHash
    }

    /// Resolve conflict with another state
    public mutating func resolveConflictWith(
        _ otherState: ComponentState,
        strategy: ComponentConflictResolutionStrategy = .newestWins
    ) {
        switch strategy {
        case .newestWins:
            if otherState.timestamp > timestamp {
                state = otherState.state
                version = otherState.version
            }
        case .highestVersionWins:
            if otherState.version > version {
                state = otherState.state
                version = otherState.version
            }
        case .merge:
            mergeState(otherState.state, strategy: .merge)
        case .manual:
            // Manual resolution required - emit conflict event
            break
        }
    }
}

// MARK: - Supporting Types

/// Merge strategy for state updates
public enum MergeStrategy: String, Codable, CaseIterable {
    case overwrite
    case keepExisting
    case merge
}

/// Conflict resolution strategy for component state
public enum ComponentConflictResolutionStrategy: String, Codable, CaseIterable {
    case newestWins
    case highestVersionWins
    case merge
    case manual
}

/// Component state metadata
public struct ComponentStateMetadata: Codable, Equatable, Hashable {
    public let lastSyncTimestamp: Date?
    public let syncAttempts: Int
    public let conflictCount: Int
    public let customData: [String: String]

    public init(
        lastSyncTimestamp: Date? = nil,
        syncAttempts: Int = 0,
        conflictCount: Int = 0,
        customData: [String: String] = [:]
    ) {
        self.lastSyncTimestamp = lastSyncTimestamp
        self.syncAttempts = syncAttempts
        self.conflictCount = conflictCount
        self.customData = customData
    }
}

/// Component state sync request
public struct ComponentStateSyncRequest: Codable {
    public let userId: String
    public let sessionId: String
    public let platform: String
    public let componentStates: [ComponentState]
    public let timestamp: Date
    public let syncVersion: Int

    public init(
        userId: String,
        sessionId: String,
        platform: String,
        componentStates: [ComponentState],
        timestamp: Date,
        syncVersion: Int
    ) {
        self.userId = userId
        self.sessionId = sessionId
        self.platform = platform
        self.componentStates = componentStates
        self.timestamp = timestamp
        self.syncVersion = syncVersion
    }
}

/// Component state sync response
public struct ComponentStateSyncResponse: Codable {
    public let success: Bool
    public let syncVersion: Int
    public let conflictsResolved: [String]
    public let timestamp: Date

    public init(
        success: Bool,
        syncVersion: Int,
        conflictsResolved: [String] = [],
        timestamp: Date = Date()
    ) {
        self.success = success
        self.syncVersion = syncVersion
        self.conflictsResolved = conflictsResolved
        self.timestamp = timestamp
    }
}

// MARK: - Codable Conformance

extension ComponentState {
    private enum CodingKeys: String, CodingKey {
        case id, componentId, instanceId, userId, sessionId
        case timestamp, version, platform, metadata
        // state requires custom encoding/decoding
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        componentId = try container.decode(String.self, forKey: .componentId)
        instanceId = try container.decode(String.self, forKey: .instanceId)
        userId = try container.decode(String.self, forKey: .userId)
        sessionId = try container.decode(String.self, forKey: .sessionId)
        timestamp = try container.decode(Date.self, forKey: .timestamp)
        version = try container.decode(Int.self, forKey: .version)
        platform = try container.decode(String.self, forKey: .platform)
        metadata = try container.decode(ComponentStateMetadata.self, forKey: .metadata)
        state = [:] // Simplified for demo
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(componentId, forKey: .componentId)
        try container.encode(instanceId, forKey: .instanceId)
        try container.encode(userId, forKey: .userId)
        try container.encode(sessionId, forKey: .sessionId)
        try container.encode(timestamp, forKey: .timestamp)
        try container.encode(version, forKey: .version)
        try container.encode(platform, forKey: .platform)
        try container.encode(metadata, forKey: .metadata)
        // state encoding simplified for demo
    }
}

// MARK: - Default Component States

extension ComponentState {

    /// Create default component state for chat message
    public static func chatMessageState(
        instanceId: String,
        userId: String,
        sessionId: String,
        isRead: Bool = false,
        isSelected: Bool = false
    ) -> ComponentState {
        return ComponentState(
            componentId: "chat-message",
            instanceId: instanceId,
            userId: userId,
            sessionId: sessionId,
            state: [
                "isRead": isRead,
                "isSelected": isSelected,
                "timestamp": Date()
            ],
            platform: "ios"
        )
    }

    /// Create default component state for user avatar
    public static func userAvatarState(
        instanceId: String,
        userId: String,
        sessionId: String,
        isOnline: Bool = false,
        lastSeen: Date? = nil
    ) -> ComponentState {
        var state: [String: Any] = [
            "isOnline": isOnline
        ]

        if let lastSeen = lastSeen {
            state["lastSeen"] = lastSeen
        }

        return ComponentState(
            componentId: "user-avatar",
            instanceId: instanceId,
            userId: userId,
            sessionId: sessionId,
            state: state,
            platform: "ios"
        )
    }

    /// Create default component state for navigation tab
    public static func navigationTabState(
        instanceId: String,
        userId: String,
        sessionId: String,
        isActive: Bool = false,
        hasNotification: Bool = false
    ) -> ComponentState {
        return ComponentState(
            componentId: "navigation-tab",
            instanceId: instanceId,
            userId: userId,
            sessionId: sessionId,
            state: [
                "isActive": isActive,
                "hasNotification": hasNotification
            ],
            platform: "ios"
        )
    }
}

File: apps/mobile/android/app/src/main/java/com/tchat/models/ComponentState.kt
package com.tchat.models

import kotlinx.serialization.Serializable
import kotlinx.serialization.Contextual
import java.util.Date
import java.util.UUID

/**
 * Component state entity for UI component state management and synchronization
 */
@Serializable
data class ComponentState(
    val id: String = UUID.randomUUID().toString(),
    val componentId: String,
    val instanceId: String,
    val userId: String,
    val sessionId: String,
    val state: MutableMap<String, @Contextual Any>,
    val timestamp: Long = System.currentTimeMillis(),
    var version: Int = 1,
    val platform: String = "android",
    val metadata: ComponentStateMetadata = ComponentStateMetadata()
) {

    /**
     * Check if state is synchronized
     */
    val isSynchronized: Boolean
        get() {
            val lastSync = metadata.lastSyncTimestamp ?: return false
            return (System.currentTimeMillis() - lastSync) < 10_000 // 10 seconds
        }

    /**
     * Get state keys
     */
    val stateKeys: Set<String>
        get() = state.keys

    /**
     * Check if state is empty
     */
    val isEmpty: Boolean
        get() = state.isEmpty()

    /**
     * Get state size (number of properties)
     */
    val stateSize: Int
        get() = state.size

    /**
     * Calculate state hash for conflict detection
     */
    val stateHash: String
        get() {
            val sortedKeys = state.keys.sorted()
            val stateString = sortedKeys.joinToString("|") { "$it:${state[it] ?: ""}" }
            return stateString.hashCode().toString()
        }

    /**
     * Update specific state property
     */
    fun updateProperty(key: String, value: Any) {
        state[key] = value
        incrementVersion()
    }

    /**
     * Update multiple state properties
     */
    fun updateProperties(updates: Map<String, Any>) {
        state.putAll(updates)
        incrementVersion()
    }

    /**
     * Remove state property
     */
    fun removeProperty(key: String) {
        state.remove(key)
        incrementVersion()
    }

    /**
     * Clear all state
     */
    fun clearState() {
        state.clear()
        incrementVersion()
    }

    /**
     * Merge state with another state
     */
    fun mergeState(otherState: Map<String, Any>, strategy: MergeStrategy = MergeStrategy.OVERWRITE) {
        when (strategy) {
            MergeStrategy.OVERWRITE -> {
                state.putAll(otherState)
            }
            MergeStrategy.KEEP_EXISTING -> {
                for ((key, value) in otherState) {
                    if (!state.containsKey(key)) {
                        state[key] = value
                    }
                }
            }
            MergeStrategy.MERGE -> {
                state.putAll(otherState)
            }
        }
        incrementVersion()
    }

    /**
     * Get state property with type safety
     */
    inline fun <reified T> getProperty(key: String): T? {
        return state[key] as? T
    }

    /**
     * Get state property with default value
     */
    inline fun <reified T> getProperty(key: String, defaultValue: T): T {
        return state[key] as? T ?: defaultValue
    }

    /**
     * Create state sync request
     */
    fun createSyncRequest(): ComponentStateSyncRequest {
        return ComponentStateSyncRequest(
            userId = userId,
            sessionId = sessionId,
            platform = platform,
            componentStates = listOf(this),
            timestamp = System.currentTimeMillis(),
            syncVersion = version
        )
    }

    /**
     * Apply sync response
     */
    fun applySyncResponse(response: ComponentStateSyncResponse): ComponentState {
        return if (response.success) {
            copy(version = response.syncVersion).markSynchronized()
        } else {
            this
        }
    }

    /**
     * Mark state as synchronized
     */
    fun markSynchronized(): ComponentState {
        return copy(
            metadata = metadata.copy(
                lastSyncTimestamp = System.currentTimeMillis()
            )
        )
    }

    /**
     * Increment version
     */
    private fun incrementVersion() {
        version++
    }

    /**
     * Detect conflicts with another state
     */
    fun hasConflictWith(otherState: ComponentState): Boolean {
        return componentId == otherState.componentId &&
                instanceId == otherState.instanceId &&
                version != otherState.version &&
                stateHash != otherState.stateHash
    }

    /**
     * Resolve conflict with another state
     */
    fun resolveConflictWith(
        otherState: ComponentState,
        strategy: ConflictResolutionStrategy = ConflictResolutionStrategy.NEWEST_WINS
    ): ComponentState {
        return when (strategy) {
            ConflictResolutionStrategy.NEWEST_WINS -> {
                if (otherState.timestamp > timestamp) {
                    copy(
                        state = otherState.state.toMutableMap(),
                        version = otherState.version
                    )
                } else {
                    this
                }
            }
            ConflictResolutionStrategy.HIGHEST_VERSION_WINS -> {
                if (otherState.version > version) {
                    copy(
                        state = otherState.state.toMutableMap(),
                        version = otherState.version
                    )
                } else {
                    this
                }
            }
            ConflictResolutionStrategy.MERGE -> {
                val mergedState = state.toMutableMap()
                mergedState.putAll(otherState.state)
                copy(
                    state = mergedState,
                    version = maxOf(version, otherState.version) + 1
                )
            }
            ConflictResolutionStrategy.MANUAL -> {
                // Manual resolution required - emit conflict event
                this
            }
            ConflictResolutionStrategy.CLIENT_WINS -> {
                this
            }
            ConflictResolutionStrategy.SERVER_WINS -> {
                copy(
                    state = otherState.state.toMutableMap(),
                    version = otherState.version
                )
            }
            ConflictResolutionStrategy.PROMPT -> {
                // Prompt user for resolution
                this
            }
        }
    }

    companion object {
        /**
         * Create default component state for chat message
         */
        fun chatMessageState(
            instanceId: String,
            userId: String,
            sessionId: String,
            isRead: Boolean = false,
            isSelected: Boolean = false
        ): ComponentState {
            return ComponentState(
                componentId = "chat-message",
                instanceId = instanceId,
                userId = userId,
                sessionId = sessionId,
                state = mutableMapOf(
                    "isRead" to isRead,
                    "isSelected" to isSelected,
                    "timestamp" to System.currentTimeMillis()
                ),
                platform = "android"
            )
        }

        /**
         * Create default component state for user avatar
         */
        fun userAvatarState(
            instanceId: String,
            userId: String,
            sessionId: String,
            isOnline: Boolean = false,
            lastSeen: Long? = null
        ): ComponentState {
            val state = mutableMapOf<String, Any>(
                "isOnline" to isOnline
            )

            lastSeen?.let { state["lastSeen"] = it }

            return ComponentState(
                componentId = "user-avatar",
                instanceId = instanceId,
                userId = userId,
                sessionId = sessionId,
                state = state,
                platform = "android"
            )
        }

        /**
         * Create default component state for navigation tab
         */
        fun navigationTabState(
            instanceId: String,
            userId: String,
            sessionId: String,
            isActive: Boolean = false,
            hasNotification: Boolean = false
        ): ComponentState {
            return ComponentState(
                componentId = "navigation-tab",
                instanceId = instanceId,
                userId = userId,
                sessionId = sessionId,
                state = mutableMapOf(
                    "isActive" to isActive,
                    "hasNotification" to hasNotification
                ),
                platform = "android"
            )
        }
    }
}

/**
 * Merge strategy for state updates
 */
@Serializable
enum class MergeStrategy {
    OVERWRITE,
    KEEP_EXISTING,
    MERGE
}


/**
 * Component state metadata
 */
@Serializable
data class ComponentStateMetadata(
    val lastSyncTimestamp: Long? = null,
    val syncAttempts: Int = 0,
    val conflictCount: Int = 0,
    val customData: Map<String, String> = emptyMap()
)

/**
 * Component state sync request
 */
@Serializable
data class ComponentStateSyncRequest(
    val userId: String,
    val sessionId: String,
    val platform: String,
    val componentStates: List<ComponentState>,
    val timestamp: Long,
    val syncVersion: Int
)

/**
 * Component state sync response
 */
@Serializable
data class ComponentStateSyncResponse(
    val success: Boolean,
    val syncVersion: Int,
    val conflictsResolved: List<String> = emptyList(),
    val timestamp: Long = System.currentTimeMillis()
)